controller/internal/controller/service_account_controller.go

```diff
@@ -0,0 +1,267 @@
+// and syncs it with the Console API state.
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
+	logger := log.FromContext(ctx)
+
+	sa := new(v1alpha1.ServiceAccount)
+	if err := r.Get(ctx, req.NamespacedName, sa); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	scope, err := NewServiceAccountScope(ctx, r.Client, sa)
+	if err != nil {
+		logger.Error(err, "failed to create scope")
+		utils.MarkCondition(sa.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Always patch object when exiting this function, so we can persist any object changes.
+	defer func() {
+		if err := scope.PatchObject(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Mark resource as not ready. This will be overridden in the end.
+	utils.MarkCondition(sa.SetCondition, v1alpha1.ReadyConditionType, v1.ConditionFalse, v1alpha1.ReadyConditionReason, "")
+
+	// Check if resource already exists in the API and only sync the ID
+	exists, err := r.isAlreadyExists(ctx, sa)
+	if err != nil {
+		utils.MarkCondition(sa.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
+		return ctrl.Result{}, err
+	}
+	if exists {
+		logger.V(9).Info("ServiceAccount already exists in the API, running in read-only mode")
+		utils.MarkCondition(sa.SetCondition, v1alpha1.ReadonlyConditionType, v1.ConditionTrue, v1alpha1.ReadonlyConditionReason, v1alpha1.ReadonlyTrueConditionMessage.String())
+		return r.handleExistingServiceAccount(ctx, sa)
+	}
+
+	// Mark resource as managed by this operator.
+	utils.MarkCondition(sa.SetCondition, v1alpha1.ReadonlyConditionType, v1.ConditionFalse, v1alpha1.ReadonlyConditionReason, "")
+
+	// Get ServiceAccount SHA that can be saved back in the status to check for changes
+	changed, sha, err := sa.Diff(utils.HashObject)
+	if err != nil {
+		logger.Error(err, "unable to calculate sa SHA")
+		utils.MarkCondition(sa.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	// Mark token as not ready if found any changes in the resource
+	if changed {
+		utils.MarkCondition(sa.SetCondition, v1alpha1.ReadyTokenConditionType, v1.ConditionFalse, v1alpha1.ReadyTokenConditionReasonError, "token not synchronized yet")
+	}
+
+	// Sync ServiceAccount CRD with the Console API
+	apiServiceAccount, err := r.sync(ctx, sa, changed)
+	if err != nil {
+		logger.Error(err, "unable to create or update sa")
+		utils.MarkCondition(sa.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	sa.Status.ID = &apiServiceAccount.ID
+	sa.Status.SHA = &sha
+
+	err = r.syncToken(ctx, sa)
```

Comment:
will this logic create a new access token on each reconcile?

Reply:
No, it checks if `ServiceAccount` has `ReadyTokenCondition` set to `False` and if that is the case then it exits early from the `syncToken` method. It will only execute if something changed (see lines 96-98).

Reply:
Ah ok